### DISCUSSION TITLE
Use WRY for prerender purposes possible?

### DISCUSSION THREAD

**Discussion by marklonquist** [tauri-apps/wry#761] [General]
Hi

Is it possible to use Wry to fetch completed html and return the html?

We're currently using prerender (https://github.com/prerender/prerender), which uses headless Chrome - and we all know how bad Chrome actually is when it comes to resource usage.

What I'm thinking is; make a rust application that exposes an API that accepts an URL (the page we want to prerender), run it through the WebView and then get back the html and return it through the API.

**Comment by amrbashir** [comment]
This is really nice use-case and I think WRY already supports that, just point it to the desired URL or load your HTML/CSS/JS directly, wait for `DOMContentLoaded` event or a timeout if you want and just use the IPC to send the value of `document.documentElement.innerHTML`

**Reply by marklonquist** [reply]
Thanks - ended up asking in discord as well. 😄

Got it somewhat working. Just need to figure out a way to wait for SPA sites to finish loading everything. 

And then check if it’s possible to do in a docker container.

**Reply by marklonquist** [reply]
OK so I've reached a point now where I'm hitting a wall.

I want to be able to make the webview render the site and then get back the HTML. But I need to do it without using the event_loop.
So basically: setup the webview, render it, return html, end.

The target is to make an API endpoint: eg. something.com/?render=https://www.google.com, which should then render the html and return it so I can cache it and then serve that to bots.

Is this something WRY can handle?

**Reply by marklonquist** [reply]
Hi @amrbashir 

Any ideas if the above is possible?

**Reply by FabianLars** [reply]
There (currently?) is no way to use a wry webview without an event_loop afaik. Did you try using `event_loop.run_return` instead of `event_loop.run`? Or what exactly is your issue?

**Reply by amrbashir** [reply]
@marklonquist OMG I am so sorry, I actually thought I wrote back. 

You have two solutions: 
1. As Fabian said, store `EventLopp` in a static variable and use `EventLoop::run_return` to advance the loop whenever you want until you get your result then stop it.
2. (I recommend this) make a small binary with WRY that takes the url as an argument, renders it and then returns the HTML through stdout or other means of communication (Sockets or just writes it to a file directly).

**Reply by ilyazub** [reply]
@marklonquist 

> So basically: setup the webview, render it, return html, end.

Probably `DOMContentLoaded` and IPC are still needed. `render it` means multiple things and `DOMContentLoaded` is one of the ways to understand when page finally rendered. `waitUntil: "load"` is needed in Puppeteer/Playwright for the same reason.

I'm working on a similar problem you've described. Will share the code once it'll be ready.

**Reply by ilyazub** [reply]
@amrbashir Thank you!

> (I recommend this) make a small binary with WRY that takes the url as an argument, renders it and then returns the HTML through stdout or other means of communication (Sockets or just writes it to a file directly).

How to get the HTML after rendering? Only by using IPC handler?

**Reply by amrbashir** [reply]
yeah (or use WebSockets but really not necessary, IPC is enough).

**Reply by ilyazub** [reply]
Here's a basic script to render HTML and print it to stdout after `window.onload`.

```rust
use std::fs::{canonicalize, read_to_string};

use wry::{
    application::{
        event::{Event, WindowEvent},
        event_loop::{ControlFlow, EventLoop},
        window::{Window, WindowBuilder},
    },
    webview::WebViewBuilder,
};

const RETURN_OUTER_HTML: &str = r#"
window.onload = function() {
  let html = document.documentElement.outerHTML;
  window.ipc.postMessage(html);
}
"#;

fn main() -> wry::Result<()> {
    enum UserEvents {
        CloseWindow,
    }

    let event_loop = EventLoop::<UserEvents>::with_user_event();

    let window = WindowBuilder::new()
        .with_title("Render HTML to string")
        .with_visible(false)
        .build(&event_loop)?;

    let html = read_to_string(canonicalize("./MY_HTML_FILE.html")?)?;

    let proxy = event_loop.create_proxy();

    let handler = move |_window: &Window, html: String| {
        println!("{}", &html);

        let _ = proxy.send_event(UserEvents::CloseWindow);
    };

    let _webview = WebViewBuilder::new(window)?
        .with_html(&html)?
        .with_initialization_script(RETURN_OUTER_HTML)
        .with_ipc_handler(handler)
        .build()?;

    event_loop.run(move |event, _, control_flow| {
        *control_flow = ControlFlow::Wait;

        match event {
            Event::WindowEvent {
                event: WindowEvent::CloseRequested,
                ..
            }
            | Event::UserEvent(UserEvents::CloseWindow) => {
                *control_flow = ControlFlow::Exit
            }
            _ => (),
        }
    });
}
```

Compile.

```bash
cargo build -r
```

It takes about one second to render HTML with JS and return back the HTML string. Only started checking the profiler.

```
time ./target/release/wry_test_render
0.17user 0.07system 0:00.81elapsed 30%CPU (0avgtext+0avgdata 80388maxresident)k
0inputs+0outputs (172major+7466minor)pagefaults 0swaps
```

Double-checking - source and resulting files are different.

```
du -sh source.html rendered.html 
688K	source.html
752K	rendered.html
```

**Reply by HaronK** [reply]
Interesting solution. Thanks.
Question to the authors of the library: is there a way to make it a WebView method (i.e. load_html) so I can call it instead of WebView::run and avoid those bunch of code above?
I can of course convert it into the function on my side but probably some other users would like to do the same.

**Reply by FabianLars** [reply]
not one of the core maintainers but this kind of helper function sounds like it's out of scope for wry to me.

**Reply by amrbashir** [reply]
I agree with @FabianLars it is out of wry scope.

**Reply by HaronK** [reply]
@amrbashir, @FabianLars Thanks for the answer.
I understand your reasons, it's just pity that it seems wry is the only cross platform library that can load dynamic pages (correct me if I'm wrong). Such a helper function would be nice to have ;)
In any case, thank you. Going to create my own wrapper.

**Comment by ilyazub** [comment]
Here's an updated code supporting wry v0.34.

```rust
use serde::{Deserialize, Serialize};
use serde_json;
use std::env;
use std::fs::{canonicalize, read_to_string};
use std::process::exit;
use wry::{
    application::{
        event::{Event, WindowEvent},
        event_loop::{ControlFlow, EventLoop},
        window::{Window, WindowBuilder},
    },
    webview::WebViewBuilder,
};

#[derive(Debug, Serialize, Deserialize)]
struct MessageParameters {
    message: String,
    data: String,
}

const RETURN_OUTER_HTML: &str = r#"
document.addEventListener("DOMContentLoaded", () => {
    let html = document.documentElement.outerHTML
    const message = JSON.stringify({ message: "dom-loaded", data: html })
    ipc.postMessage(message)
})
"#;

fn main() -> wry::Result<()> {
    let event_loop = EventLoop::new();

    let window = WindowBuilder::new()
        .with_visible(false)
        .build(&event_loop)?;

    let args: Vec<String> = env::args().collect();
    let file_path = &args[1];
    let html = read_to_string(canonicalize(file_path)?)?;

    let handler = move |_window: &Window, req: String| {
        let msg: MessageParameters = serde_json::from_str(&req).unwrap();

        if &msg.message == "dom-loaded" {
            println!("{}", &msg.data);
        }

        exit(0);
    };

    let _webview = Some(
        WebViewBuilder::new(window)?
            .with_html(&html)?
            .with_initialization_script(RETURN_OUTER_HTML)
            .with_ipc_handler(handler)
            .build()?,
    );

    event_loop.run(move |event, _, control_flow| {
        *control_flow = ControlFlow::Wait;

        match event {
            Event::WindowEvent {
                event: WindowEvent::CloseRequested,
                ..
            } => *control_flow = ControlFlow::Exit,

            _ => (),
        }
    })
}
```

Let's check how it works, using the Google Shopping results page (the HTML file is taken via the our [Google Shopping Results scraper](https://serpapi.com/google-shopping-api)).

```bash
$ time cargo run -r ~/serpapi/tmp/big_shopping.html 2>/dev/null > ~/serpapi/tmp/big_shopping_rendered.html

real	0m1.193s
user	0m0.360s
sys	0m0.199s
```

The source and resulting HTML files differ.

```bash
$ du -sh ~/serpapi/tmp/big_shopping*.html
1.9M	/home/serpapi/tmp/big_shopping.html
2.4M	/home/serpapi/tmp/big_shopping_rendered.html
```

Comparing actual HTML.

![image](https://github.com/tauri-apps/wry/assets/282605/9389766a-b5ec-4a61-af9c-422e60f79f3f)

In the original HTML (on the left), thumbnails are 1x1 placeholder pixels &mdash; `data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==`.

In the resulting HTML (on the right), thumbnails are actual images, that are JavaScript-rendered.

**Closing thoughts**

1. Now, it's interesting to compare the JS execution performance of wry and [`webview`](https://github.com/webview/webview)-based renderers ([`webview_deno`](https://github.com/webview/webview_deno) or [`webview_ruby`](https://github.com/Maaarcocr/webview_ruby)).

2. It's even more interesting to bundle a HTML prerenderer into a single binary with bundled shared libraries so it can be portable.

**Comment by marklonquist** [comment]
I still think WRY would be perfect for helping with "prerender" tasks. Current solutions involve using puppeteer and actually opening up a browser, navigate to the page, wait for navigation to finish (among other things), and then getting the content and returning it to the bots.

Seems it would perform way better with WRY if we can get it working correctly (eg. waiting for JS to finish loading completely)

**Comment by ilyazub** [comment]
> Current solutions involve using puppeteer and actually opening up a browser, navigate to the page, wait for navigation to finish (among other things), and then getting the content and returning it to the bots.

The same happens with wry, except it uses webkit instead of chromium.

**Reply by marklonquist** [reply]
we should also remember, that performance is in play here, and from my local tests Wry seems faster than using chromium

**Reply by ilyazub** [reply]
> we should also remember, that performance is in play here, and from my local tests Wry seems faster than using chromium

That's interesting. May you share the benchmark?

(I didn't notice the difference in the startup, JS execution, and rendering time, when compared wry (webkit) vs puppeteer (chromium).)
